### DISCUSSION TITLE
Form tamper protection

### DISCUSSION THREAD

**Discussion by SamMoon1978** [adonisjs/core#3525] [Help]
Been lurking around here forever and found answers to all my questions but this one.
Is there any built in tamper protection in your framework for forms on post?

Such as validation in Laravel or tamper protection in CakePHP?

**Comment by Julien-R44** [comment]
Hi, I'm not sure what you mean by tamper protection but I guess this is what you are looking for: https://docs.adonisjs.com/guides/validator/introduction

**Reply by SamMoon1978** [reply]
Thank you Julien, not exactly what I am looking for, maybe my example question was a little too vague.

Copied some code below to illustrate better.




Controller:

```
export default class SampleController {
  //Select item to edit and render form
 public async editItem({ request, response, view, session, bouncer }: HttpContextContract) {
    await bouncer.authorize('createOrEditItem')
    const item = await Item
                        .query()
                        .whereRaw('id =?', Encryption.decrypt(request.param('encrypted_item_id')))
                       .andWhereRaw('user_id =?', session.get('user_id'))`
                        .first()`
    return view.render('edititem', { item })
 }
}
```
The above renders a post form sending the values of the form to the following controller:

```
  public async saveItem({ view, request, response, session, bouncer }: HttpContextContract) {
    await bouncer.authorize('createOrEditItem')
    //Check if submitted item is the same as one from edit
    const item = await Item
                        .query()
                        .whereRaw('id = ?', Encryption.decrypt(request.input('encrypted_item_id')))
                        .andWhereRaw('user_id = ?', session.get('user_id'))
                        .first()
 }
```

I am already ensuring that whoever does the edit has the right to edit via a Bouncer and also ensure he/she can only edit items authorized to edit.

However if I have two tabs open editing two different items and copy the encrypted_item_id from tab 1 to tab 2 via inspecting the form and change the hidden field that I am using to know which item I want to update it overwrites item 2 with data from item 1.

CakePHP and Laravel have build in functionality to create hashes for the entire form that i can then compare and also support storing original values in session to compare against.

I can do this manually but was hoping there is a library or out of the box functionality to achieve this.

Does this make sense?
